specs/governance/gov-token.md

```diff
@@ -24,72 +24,74 @@
 
 `GovernanceToken` is an [ERC20](https://eips.ethereum.org/EIPS/eip-20) token contract that inherits from `ERC20Burnable`,
 `ERC20Votes`, and `Ownable`. It allows token holders to delegate their voting power to other addresses, enabling a representative
-voting system.
+voting system. The contract integrates with the `Alligator` contract through a hook-based approach to enable advanced delegation
+functionality, including partial delegation and subdelegations.
 
-The specific implementation details of the voting and delegation logic are inherited from the `ERC20Votes` contract. The
-`GovernanceToken` contract focuses on integrating this functionality with `ERC20Burnable` and adding the minting
-capability. EIP-712 permit functionality is also inherited from `ERC20Permit`, which is imported by `ERC20Votes`.
+### Hook-based Integration with Alligator
 
-### Checkpoints
+The `GovernanceToken` contract integrates with the `Alligator` contract through a hook-based approach to enable advanced
+delegation features. The `_afterTokenTransfer` function in the `GovernanceToken` is modified to call the `afterTokenTransfer`
+function in the `Alligator` contract, allowing the `Alligator` contract to consume the hooks and update its delegation and
+checkpoint mappings accordingly.
 
-`ERC20Votes` uses checkpoints to track voting power history. A checkpoint is a "snapshot" of the voting power of an
-address at a specific block number.
+If the call to the Alligator's `afterTokenTransfer` function fails, the token transfer must still be completed, and the following
+`AlligatorCallFailed` event will be emitted to indicate the failed call:
 
 ```solidity
-struct Checkpoint {
-    uint32 fromBlock;
-    uint224 votes;
-}
+event AlligatorCallFailed(address indexed transferFrom, address indexed transferTo, uint256 amount);
 ```
 
-Checkpoints are organized as a one-to-one mapping of user addresses to append-only lists of checkpoints. When a
-user is involved in a token transfer, mint, burn, or vote delegation, a new checkpoint is added to their corresponding
-list of checkpoints, reflecting the user's updated voting power at that block number.
+This prevents the token transfer from being blocked due to issues with the Alligator contract.
```

Comment:
What happens to delegations in this case? It feels like this would result in broken invariants. For example, Alice has 100 tokens and delegates them all to Bob. Alice transfers all 100 tokens to Clara, but the call to Alligator fails. Now, Clara has 100 tokens and no voting weight, and Bob has 100 votes he's not entitled to, because Clara should have those votes.

My initial reaction to this problem is that instead the transfer should revert if the call reverts, to ensure the token holders always have voting power. If calls to Alligator are reverting, that is a high severity bug that we would prioritize deploying a fix for

Reply:
implemented! feel free to take a look:


https://github.com/ethereum-optimism/specs/pull/233/files#diff-e1c4b822c42f7fba380f47d85e0f5636325095de27c085d526bd8bb4429f1dc3R36-R38

Reply:
Looks great, thank you!